src/Microsoft.DotNet.Interactive.Js/src/dotnet-interactive/client-side-kernel.ts

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { ClientSideKernelInvocationContext } from "./client-side-kernel-invocation-context";
+import { DisposableSubscription, KernelEventEnvelopeObserver, KernelCommandEnvelope, KernelTransport, Disposable, KernelEventEnvelope } from "./contracts";
+import { Kernel, KernelInvocationContext } from "./dotnet-interactive-interfaces";
+import { TokenGenerator } from "./tokenGenerator";
+
+export class ClientSideKernel implements Kernel {
+    private _commandHandlers: { [commandType: string]: (envelope: KernelCommandEnvelope, context: KernelInvocationContext) => Promise<void> } = {};
+    private readonly _eventSubscribers: { [token: string]: KernelEventEnvelopeObserver} = {};
+    private readonly _tokenGenerator: TokenGenerator = new TokenGenerator();
+
+    // Is it worth us going to efforts to ensure that the Promise returned here accurately reflects
+    // the command's progress? The only thing that actually calls this is the kernel transport, through
+    // the callback set up by attachKernelToTransport, and the callback is expected to return void, so
+    // nothing is ever going to look at the promise we return here.
+    send(command: KernelCommandEnvelope): Promise<void> {
+
+        let handler = this._commandHandlers[command.commandType];
+        if (handler) {
+            let resolvePromise: () => void;
+            let promise = new Promise<void>(r => resolvePromise = r);
+                let _: Promise<void> = (async () => {
+                let context = ClientSideKernelInvocationContext.establish(command);
+
+                let isRootCommand = context.command === command;
+                let contextEventsSubscription: Disposable = null;
+
+                if (isRootCommand) {
+                    contextEventsSubscription = context.subscribeToKernelEvents(e => this.publishEvent(e));
+                }
+
+                await handler(command, context);
+
+                if (isRootCommand) {
+                    context.dispose();
+                } else {
+                    context.complete(command);
+                }
+
+                if (contextEventsSubscription) {
+                    contextEventsSubscription.dispose();
+                }
+
+                resolvePromise();
+            })();
+            return promise;
+
+        } else {
+            // TODO: what's the right behaviour if there is no handler?
```

Comment:
I think for the moment is fine to reqject the promise. In the .NET counterpart a handler is always set

Reply:
If no handler is set in the .NET implementation, I believe it will always coincide with a `NoSuitableKernelException`.